src/audio/kpb.c

```diff
@@ -1593,6 +1594,70 @@ static enum task_state kpb_draining_task(void *arg)
 	return SOF_TASK_STATE_COMPLETED;
 }
 
+#if KPB_HIFI4
+void kpb_convert_24b_to_32b(const void *linear_source, int ioffset,
+			    struct audio_stream __sparse_cache *sink, int ooffset,
+			    unsigned int n_samples)
+{
+	int ssize = audio_stream_sample_bytes(sink);
+	uint8_t *in = (uint8_t *)linear_source + ioffset * ssize;
+	uint8_t *out = audio_stream_wrap(sink, (uint8_t *)sink->w_ptr + ooffset * ssize);
+	ae_int32x2 *out_ptr = (ae_int32x2 *)(out);
+
+	ae_valign align_in = AE_LA64_PP(in);
+	int i = 0;
+	ae_int24x2 d24 = AE_ZERO24();
+
+	if (!IS_ALIGNED((uintptr_t)out_ptr, 8)) {
+		AE_LA24_IP(d24, align_in, in);
+		ae_int32x2 d320 = d24;
+		int higher = AE_MOVAD32_H(d320);
+		*(ae_int32 *)(out_ptr) = higher << 8;
+		out_ptr = (ae_int32x2 *)(out + 4);
+		++i;
+	}
+	// process two samples in single iteration to increase performance
+	while (i < (int)n_samples - 1) {
+		AE_LA24X2_IP(d24, align_in, in);
+		ae_int32x2 d320 = d24;
+
+		d320 = AE_SLAI32(d320, 8);
+		AE_S32X2_IP(d320, out_ptr, 8);
+		i += 2;
+	}
+	if (i != (int)n_samples) {
+		AE_LA24X2_IP(d24, align_in, in);
+		ae_int32x2 d320 = d24;
+		int higher = AE_MOVAD32_H(d320);
+		*(ae_int32 *)(out_ptr) = higher << 8;
+	}
+}
+#else
+void kpb_convert_24b_to_32b(const void *source, int ioffset,
+			    struct audio_stream __sparse_cache *sink,
+			    int ooffset, unsigned int samples)
+{
+	comp_cl_info(&comp_kpb, "kpb_pcm_convert_s24_c24_to_s24_c32() entry, samples %d", samples);
+	int ssize = audio_stream_sample_bytes(sink);
+	uint8_t *src = (uint8_t *)source + ioffset * 3;
+	int32_t *dst = audio_stream_wrap(sink, (int32_t *)sink->w_ptr + ooffset * ssize);
+	int processed;
+	int nmax, i, n;
+
+	for (processed = 0; processed < samples; processed += n) {
+		dst = audio_stream_wrap(sink, dst);
+		n = samples - processed;
+		nmax = audio_stream_bytes_without_wrap(sink, dst) >> BYTES_TO_S32_SAMPLES;
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += 1) {
+			*dst = (*(src + 2) << 24) | (*(src + 1) << 16) | (*(src + 0) << 8);
+			*dst >>= 8;
```

Comment:
can we just make this
```
*dst = (src[2] << 16) | (src[1] << 8) | src[0];
```

Reply:
With the new approach this line should be removed:
`*dst >>= 8;`